lib/zip/file.rb

```diff
@@ -405,14 +405,13 @@ def check_file(path)
     def on_success_replace
       tmpfile      = get_tempfile
       tmp_filename = tmpfile.path
-      ObjectSpace.undefine_finalizer(tmpfile)
-      tmpfile.close
       if yield tmp_filename
-        ::File.rename(tmp_filename, name)
+        new_file = ::File.new(name, "w")
+        IO.copy_stream tmpfile, new_file
```

Comment:
I see. The problem I am trying to get around is:

* is `tmpfile.close` guaranteed to leave the file on the filesystem, I think it is not, hence https://github.com/rubyzip/rubyzip/pull/185
* we could solve the problem from https://github.com/rubyzip/rubyzip/pull/185 and maybe other problems by moving tmpfile.close to below the rename (as I have done in 7432e834c12e8314b84ad2d475bb55e77cb90e1b), but ``::File.rename`` on most (all?) systems uses the same file with the same inode. is `tmpfile.close` guaranteed to delete by file path and not by inode? the if so, then that's the solution. looking at the MRI implementation (https://github.com/ruby/ruby/blob/trunk/lib/tempfile.rb#L159, https://github.com/ruby/ruby/blob/trunk/lib/tempfile.rb#L203), this seems to be the case, so this could be the way to go.

So, we could remove 1ad5814fd95f46fc3504306899ccfb58c7e3479c from my PR and leave the other two.

Thoughts?